File: client/src/main/java/util/SimulationResults.java
package util;

import java.text.NumberFormat;
import java.util.Vector;

public class SimulationResults
{
	private double wins = 0;
	private double perfectGames = 0;
	private double goodChallenges = 0;
	private double totalChallenges = 0;
	private double opportunitiesToChallenge = 0;
	
	public void incrementWins()
	{
		wins++;
	}
	
	public void incrementPerfectGames()
	{
		perfectGames++;
	}
	
	public void incrementGoodChallenges()
	{
		goodChallenges++;
	}
	
	public void incrementTotalChallenges()
	{
		totalChallenges++;
	}
	
	public void incrementOpportunitiesToChallenge()
	{
		opportunitiesToChallenge++;
	}
	
	public Vector<String> generateRow(int opponentNumber, String strategy, double totalGames)
	{
		String winRate = formatPercent(wins, totalGames);
		String challengeRate = formatPercent(totalChallenges, opportunitiesToChallenge);
		String challengeSuccessRate = formatPercent(goodChallenges, totalChallenges);
		String perfectRate = formatPercent(perfectGames, wins);
		
		Vector<String> row = new Vector<>();
		row.add("" + opponentNumber);
		row.add(strategy);
		row.add(winRate);
		row.add(challengeRate);
		row.add(challengeSuccessRate);
		row.add(perfectRate);
		
		return row;
	}
	
	private String formatPercent(double numerator, double denominator)
	{
		if (denominator == 0)
		{
			return "";
		}
		
		NumberFormat percentFormat = NumberFormat.getPercentInstance();
		percentFormat.setMinimumFractionDigits(2);
		return percentFormat.format(numerator/denominator);
	}
}

File: client/src/main/kotlin/util/SimulationResults.kt
package util

import java.util.Vector

class SimulationResults {
    private var wins = 0.0
    private var perfectGames = 0.0
    private var goodChallenges = 0.0
    private var totalChallenges = 0.0
    private var opportunitiesToChallenge = 0.0

    fun incrementWins() {
        wins++
    }

    fun incrementPerfectGames() {
        perfectGames++
    }

    fun incrementGoodChallenges() {
        goodChallenges++
    }

    fun incrementTotalChallenges() {
        totalChallenges++
    }

    fun incrementOpportunitiesToChallenge() {
        opportunitiesToChallenge++
    }

    fun getWinRate(totalGames: Int) = computePercentage(wins, totalGames.toDouble())

    fun getChallengeRate() = computePercentage(totalChallenges, opportunitiesToChallenge)

    fun getChallengeSuccessRate() = computePercentage(goodChallenges, totalChallenges)

    fun getPerfectRate() = computePercentage(perfectGames, wins)

    private fun computePercentage(numerator: Double, denominator: Double): Double? {
        if (denominator == 0.0) {
            return null
        }

        return MathsUtil.getPercentage(numerator, denominator, 2)
    }

    fun generateRow(opponentNumber: Int, strategy: String, totalGames: Int): Vector<String> {
        val row = Vector<String>()
        row.add("" + opponentNumber)
        row.add(strategy)
        row.add("" + getWinRate(totalGames))
        row.add("" + getChallengeRate())
        row.add("" + getChallengeSuccessRate())
        row.add("" + getPerfectRate())

        return row
    }
}
